eureka-client/src/main/java/com/netflix/discovery/DiscoveryClient.java

```diff
@@ -1072,11 +1088,60 @@ private ClientResponse makeRemoteCall(Action action) throws Throwable {
      * @throws Throwable
      *             on any error.
      */
-    private ClientResponse makeRemoteCall(Action action, int serviceUrlIndex)
-            throws Throwable {
+    private ClientResponse makeRemoteCall(Action action, int serviceUrlIndex) throws Throwable {
+        String serviceUrl;
+        try {
+            serviceUrl = eurekaServiceUrls.get().get(serviceUrlIndex);
+            return makeRemoteCallWithFollowRedirect(action, serviceUrl);
+        } catch (Throwable t) {
+            if (eurekaServiceUrls.get().size() > ++serviceUrlIndex) {
+                logger.warn("Trying backup: " + eurekaServiceUrls.get().get(serviceUrlIndex));
+                SERVER_RETRY_COUNTER.increment();
+                return makeRemoteCall(action, serviceUrlIndex);
+            } else {
+                ALL_SERVER_FAILURE_COUNT.increment();
+                logger.error("Can't contact any eureka nodes - possibly a security group issue?", t);
+                throw t;
+            }
+        }
+    }
+
+    private ClientResponse makeRemoteCallWithFollowRedirect(Action action, String serviceUrl) throws Throwable {
+        URI targetUrl = new URI(serviceUrl);
+        for(int followRedirectCount = 0; followRedirectCount < 10; followRedirectCount++) {
+            ClientResponse clientResponse = makeRemoteCall(action, targetUrl.toString());
+            if(clientResponse.getStatus() < 300) {
+                if(followRedirectCount > 0) {
+                    Matcher pathMatcher = REDIRECT_PATH_REGEX.matcher(targetUrl.getPath());
+                    if(pathMatcher.matches()) {
```

Comment:
There shouldn't be, assuming server is in charge of making the decision
about that. I will add a header that will notify server side if a
particular client request can be redirected (by default no). Even if client
states yes, some requests can be handled directly by server (for example
full/delta fetches that we do not support on bootstrap cluster,
registrations, etc).

On Mon, Mar 30, 2015 at 4:25 PM, Nitesh Kant <notifications@github.com>
wrote:

> In eureka-client/src/main/java/com/netflix/discovery/DiscoveryClient.java
> <https://github.com/Netflix/eureka/pull/486#discussion_r27441671>:
>
> > +            } else {
> > +                ALL_SERVER_FAILURE_COUNT.increment();
> > +                logger.error("Can't contact any eureka nodes - possibly a security group issue?", t);
> > +                throw t;
> > +            }
> > +        }
> > +    }
> > +
> > +    private ClientResponse makeRemoteCallWithFollowRedirect(Action action, String serviceUrl) throws Throwable {
> > +        URI targetUrl = new URI(serviceUrl);
> > +        for(int followRedirectCount = 0; followRedirectCount < 10; followRedirectCount++) {
> > +            ClientResponse clientResponse = makeRemoteCall(action, targetUrl.toString());
> > +            if(clientResponse.getStatus() < 300) {
> > +                if(followRedirectCount > 0) {
> > +                    Matcher pathMatcher = REDIRECT_PATH_REGEX.matcher(targetUrl.getPath());
> > +                    if(pathMatcher.matches()) {
>
> Are there redirects from the server that we need to ignore?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/Netflix/eureka/pull/486/files#r27441671>.
>